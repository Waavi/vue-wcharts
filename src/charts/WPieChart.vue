<script>
import chartMixin from '../mixins/chart'

export default {
    name: 'WPieChart',
    mixins: [chartMixin],
    render (h) {
        const slots = this.$slots.default || []
        let datakeys = [] // We need get slots datakey prop to calculate max and min values for the scales
        const pies = [] // We need inject necessary props to cartesian slots
        const others = []
        const plugins = []

        // Props
        const activeElements = []
        const legends = []

        // Reset snap to manage bars
        this.snap = {}

        slots.forEach((slot) => {
            const options = slot.componentOptions
            if (!options) {
                others.push(slot)
                return
            }
            const sealed = options.Ctor.sealedOptions
            if (!sealed) {
                return
            }
            const { propsData } = options
            const { datakey, legend } = propsData
            const piesLength = pies.length

            switch (sealed.type) {
                case 'pie':
                    // Add datekeys. Removed unique datakey condiition to use multiple elements
                    if (datakey) datakeys = [...datakeys.filter(key => key !== datakey), datakey]
                    // Add to actives elements
                    activeElements.push(piesLength)
                    // Add to legends elements
                    if (legend) legends.push(legend)
                    // Add slot
                    slot.index = piesLength
                    pies.push(slot)
                    break
                case 'plugins':
                    plugins.push(slot)
                    break
                default:
                    break
            }

            if (sealed.preload) {
                sealed.preload({ parent: this, props: propsData, index: piesLength })
            }
        })

        const { viewWidth, height, responsive } = this
        this.datakeys = datakeys
        this.activeElements = activeElements
        this.legends = legends

        return h(
            'div',
            {
                class: 'WCartesian',
                style: {
                    position: 'relative',
                    width: responsive ? '100%' : `${viewWidth}px`,
                },
            },
            [
                h(
                    'svg',
                    {
                        attrs: {
                            width: viewWidth,
                            height,
                            viewBox: `0 0 ${viewWidth} ${height}`,
                        },
                    },
                    this.chartReady ? [others, pies] : []
                ),
                this.chartReady ? plugins : [],
            ]
        )
    },
}
</script>
