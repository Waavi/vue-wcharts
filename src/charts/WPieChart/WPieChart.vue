<script>
import VueTypes from 'vue-types'
import chartMixin from '../../mixins/chart'
import { getPercentValue } from '../../utils/mathsPie'

export default {
    name: 'WPieChart',
    mixins: [chartMixin],
    props: {
        cx: VueTypes.string.def('50%'),
        cy: VueTypes.string.def('50%'),
        startAngle: VueTypes.number.def(0),
        endAngle: VueTypes.number.def(360),
        paddingAngle: VueTypes.number.def(0),
    },
    computed: {
        curCx () {
            return getPercentValue({ percent: this.cx, totalValue: this.width, defaultValue: this.width / 2 })
        },
        curCy () {
            return getPercentValue({ percent: this.cy, totalValue: this.height, defaultValue: this.height / 2 })
        },
    },
    render (h) {
        const slots = this.$slots.default || []
        let datakeys = [] // We need get slots datakey prop to calculate max and min values for the scales
        const pies = [] // We need inject necessary props to cartesian slots
        const others = []
        const plugins = []

        // Props
        const legends = []

        // Reset snap to manage bars
        this.snap = {}

        slots.forEach((slot) => {
            const options = slot.componentOptions
            if (!options) {
                others.push(slot)
                return
            }
            const sealed = options.Ctor.sealedOptions
            if (!sealed) {
                return
            }
            const { propsData } = options
            const { datakey, legend } = propsData || {}
            const piesLength = pies.length

            switch (sealed.type) {
                case 'pie':
                    // Add datekeys. Removed unique datakey condiition to use multiple elements
                    if (datakey) datakeys = [...datakeys.filter(key => key !== datakey), datakey]
                    // Add to legends elements
                    if (legend) legends.push(legend)
                    // Add slot
                    slot.componentOptions.propsData.index = piesLength
                    pies.push(slot)
                    break
                case 'plugins':
                    plugins.push(slot)
                    break
                default:
                    break
            }

            if (sealed.preload) {
                sealed.preload({ parent: this, props: propsData, index: piesLength })
            }
        })

        const { viewWidth, height, responsive } = this
        this.datakeys = datakeys
        this.legends = legends

        return h(
            'div',
            {
                class: 'WPieChart',
                style: {
                    position: 'relative',
                    width: responsive ? '100%' : `${viewWidth}px`,
                },
            },
            [
                h(
                    'svg',
                    {
                        attrs: {
                            width: viewWidth,
                            height,
                            viewBox: `0 0 ${viewWidth} ${height}`,
                        },
                    },
                    this.chartReady ? [others, pies] : []
                ),
                this.chartReady ? plugins : [],
            ]
        )
    },
}
</script>
