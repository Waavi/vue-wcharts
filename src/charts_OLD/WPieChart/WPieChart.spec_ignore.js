// import { mount, shallowMount } from '@vue/test-utils'
// import localVue from '../../../config/tests'
// import WPieChart from './WPieChart.vue'

// describe('Charts/WPieChart', () => {
//     const dataset = [
//         {
//             name: 'Page A',
//             one: 1000,
//             two: 2400,
//             three: 2400,
//         },
//         {
//             name: 'Page B',
//             one: 3000,
//             two: 1398,
//             three: 2210,
//         },
//         {
//             name: 'Page C',
//             one: 2000,
//             two: 9800,
//             three: 2290,
//         },
//     ]
//     const propsData = {
//         dataset,
//     }

//     const defaultConfig = {
//         localVue,
//         propsData,
//     }

//     xit(`Should be render correctly`, () => {
//         const wrapper = shallowMount(WPieChart, defaultConfig)
//         expect(wrapper.html()).toMatchSnapshot()
//     })

//     xit(`Should be render responsive mode correctly`, () => {
//         const wrapper = shallowMount(WPieChart, {
//             ...defaultConfig,
//             propsData: {
//                 ...propsData,
//                 responsive: true,
//             },
//         })
//         expect(wrapper.html()).toMatchSnapshot()
//     })

//     xit(`Should be render correctly with WPie component and WTooltip`, () => {
//         const wrapper = mount(WPieChart, {
//             ...defaultConfig,
//             slots: {
//                 default: `
//                     <WPie datakey="one" :radius="[110, 150]" />
//                     <WTooltip />
//                 `,
//             },
//         })

//         expect(wrapper).toMatchSnapshot()
//     })

//     xit(`Should be ommit children components not allowed`, () => {
//         const wrapper = mount(WPieChart, {
//             ...defaultConfig,
//             slots: {
//                 default: `
//                     <WLine datakey="two" />
//                     <WPie datakey="one" :radius="[110, 150]" />
//                     <WTooltip />
//                 `,
//             },
//         })

//         expect(wrapper).toMatchSnapshot()
//     })
// })
