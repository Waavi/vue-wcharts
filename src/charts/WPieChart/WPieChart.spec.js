import { shallowMount } from '@vue/test-utils'
import WPieChart from './WPieChart.vue'

describe('Charts/WCartesian', () => {
    const dataset = [
        {
            name: 'Page A',
            one: 1000,
            two: 2400,
            three: 2400,
        },
        {
            name: 'Page B',
            one: 3000,
            two: 1398,
            three: 2210,
        },
        {
            name: 'Page C',
            one: 2000,
            two: 9800,
            three: 2290,
        },
    ]
    const propsData = {
        dataset,
    }

    const defaultConfig = {
        propsData,
    }
    // TODO:
    // add slots of WPie
    it(`Should be render correctly`, () => {
        const wrapper = shallowMount(WPieChart, defaultConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })
})
