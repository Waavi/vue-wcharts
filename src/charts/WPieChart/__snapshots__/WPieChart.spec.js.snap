// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Charts/WCartesian Should be render correctly 1`] = `<div class="WPieChart" style="position: relative; width: 600px;"><svg width="600" height="400" viewBox="0 0 600 400"></svg></div>`;
